lib/Bio/EnsEMBL/DataCheck/Checks/CheckFlatProteinTrees.pm

```diff
@@ -0,0 +1,122 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+
+=cut
+
+package Bio::EnsEMBL::DataCheck::Checks::CheckFlatProteinTrees;
+
+use warnings;
+use strict;
+
+use Moose;
+use Test::More;
+use Bio::EnsEMBL::DataCheck::Test::DataCheck;
+
+extends 'Bio::EnsEMBL::DataCheck::DbCheck';
+
+use constant {
+  NAME           => 'CheckFlatProteinTrees',
+  DESCRIPTION    => 'Check protein tree integrity ensuring number of leaves with parent node at root < 3',
+  GROUPS         => ['compara', 'compara_protein_trees'],
+  DATACHECK_TYPE => 'critical',
+  DB_TYPES       => ['compara'],
+  TABLES         => ['gene_tree_node', 'gene_tree_root']
+};
+
+sub tests {
+  my ($self) = @_;
+  my $dba = $self->dba;
+  my $helper = $dba->dbc->sql_helper;
+
+  # Count and collect nodes with root_id != 0 and is not a seq_member
+  my $internal_node_sql = qq/
+    SELECT gtn.root_id, COUNT(*) AS internal_nodes 
+      FROM gene_tree_node gtn 
+    WHERE gtn.seq_member_id IS NULL 
+      AND gtn.node_id <> gtn.root_id 
+      AND gtn.root_id <> 0 
+    GROUP BY gtn.root_id
+  /;
+  my $internal_node = $helper->execute_into_hash(  
+   -SQL => $internal_node_sql
+  );
+
+  # Count and collect seq_members where root_id != 0, parent_id==root_id and count>1
+  my $flat_member_sql = qq/
+    SELECT gtn.root_id, COUNT(*) AS root_members 
+      FROM gene_tree_node gtn 
+    WHERE gtn.seq_member_id IS NOT NULL 
+      AND gtn.parent_id = gtn.root_id 
+      AND gtn.root_id <> 2 
+    GROUP BY gtn.root_id 
+    HAVING root_members > 1
+  /;
+  my $flat_members = $helper->execute(  
+   -SQL => $flat_member_sql,
+   -USE_HASHREFS => 1
+  );
+
+  # Count and collect all node memmbers where root_id != 0
+  my $all_members_sql = qq/
+    SELECT root_id, COUNT(DISTINCT root_id) AS root_count 
+      FROM gene_tree_node 
+    WHERE root_id <> 0 
+    GROUP BY root_id
+  /;
+  my $all_members = $helper->execute_into_hash(  
+   -SQL => $all_members_sql
+  );
+
+  # Collect all non-rooted trees
+  my $nonrooted_trees_sql = qq/
+    SELECT root_id, 1
+      FROM gene_tree_root 
+    WHERE tree_type = 'tree' 
+      AND clusterset_id LIKE '%_it_%' 
+      AND clusterset_id NOT LIKE 'pg_it_%'
+  /;
+  my $nonrooted_trees = $helper->execute_into_hash( 
+    -SQL => $nonrooted_trees_sql
+  );
+
+  my @flat_trees;
+  my @flat_trees_w_structure;
+
+  foreach my $flat_member ( @$flat_members ) {
+    my $node_id = $flat_member->{root_id};
+    my $count = $flat_member->{root_members};
+    # Check if $node_id is present in $nonrooted_trees
+    if (exists $internal_node->{$node_id} ) {
+      if ( !exists $nonrooted_trees->{$node_id} ) {
+        push @flat_trees_w_structure, $node_id;
+      }
+    } # Check that root has no more than 2 members
+    elsif ( defined $count ) {
+      if ( ($count > 2) && ($count == $all_members->{$node_id} )) {
+        push @flat_trees, $node_id;
+      }
+    }
+  }
```

Comment:
I will think about this test a bit more, but here is my current analysis of the HC code.

I think this test is too permissive for the "nonrooted trees". I would code it like this.
1. check `$count` (number of members directly attached to the root)
   * 2 → OK
   * 3 → OK only if part of the "nonrooted trees", otherwise NO
   * 4+ -> NO
2. When NO, check whether there are internal nodes and push to `@flat_trees` or `flat_trees_w_structure`. This can be alternatively checked by comparing `$count` to `$all_members->{$root_id}`. This also means that only one of `$all_members` / `$internal_node` is needed.

I hope our current trees would pass this :crossed_fingers: